### PR TITLE
Users can gives notes a custom colour

### PR DESCRIPTION
**Pull request by altair1673**
### Prerequisites

- [X] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

Notes are by default this purple colour but now the user can use a JavaFX colour picker thats on every note to give it a custom colour.
### Why Should This Be In Core?
Gives users to differentiate between notes. Eg. More important notes could be coloured in red so the user can prioritize those.

### Benefits

Better organization of notes.
Notes View can have visually appealing colours.
### Possible Drawbacks

Since each note has a colour option the UI could look cluttered
### Verification Process

Colour of note changes depending on whatever is chosen. Notes retain colour since it is saved in state.
### Applicable Issues

#1760

### CODE REVIEW

**Review by Delphinus8821 — CHANGES_REQUESTED** [review]
A few bugs found with testing
A null pointer when opening an older graph that includes user notes:
image

Auto notes are no longer blue:
image

After editing a note, the colour picker gets put above the edit button:
image

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewEntry.java [hunk 1]
```diff
@@ -30,17 +31,19 @@ public class NotesViewEntry implements PluginReportListener {
...
+    private String nodeColour;
     private final Boolean userCreated;
     private Boolean graphAttribute;
     private List<Integer> nodesSelected;
     private List<Integer> transactionsSelected;
     private List<String> tags = new ArrayList<>();
     private boolean editMode;
 
-    public NotesViewEntry(final String dateTime, final String noteTitle, final String noteContent, final boolean userCreated, final boolean graphAttribute) {
+    public NotesViewEntry(final String dateTime, final String noteTitle, final String noteContent, final boolean userCreated, final boolean graphAttribute, String nodeColour) {
```
Comment (Delphinus8821) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewEntry.java:42]: Final on parameter

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewEntry.java [hunk 2]
```diff
@@ -102,6 +105,14 @@ public List<String> getTags() {
         return tags;
     }
 
+    public String getNodeColour() {
+        return nodeColour;
+    }
+
+    public void setNodeColour(String nodeColour) {
```
Comment (Delphinus8821) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewEntry.java:112]: Final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 3]
```diff
@@ -47,14 +48,19 @@
 import javafx.collections.FXCollections;
 import javafx.collections.ListChangeListener;
 import javafx.collections.ObservableList;
+import javafx.event.ActionEvent;
+import javafx.event.EventHandler;
```
Comment (Delphinus8821) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:52]: A lot of these imports are unused

**Review by Delphinus8821 — CHANGES_REQUESTED** [review]

**Comment by OrionsGuardian** [comment]
Null pointer when loading an old graph:
image

**Comment by OrionsGuardian** [comment]
There's a problem when re-loading a graph that has user defined coloured notes attached to specific graph elements ...
The notes are no longer attached correctly to the graph elements.
ie. a Vertex note is now linked to a transaction, and a Transaction note isn't linked to anything at all.
image

**Comment by OrionsGuardian** [comment]
Another exception:
If you edit a Note, but don't save it yet, click on another graph and click back to the edited graph, then click on the Save button, it will throw an exception.
(also fyi, the colour picker appears in the edit window in that scenario as well)
```
java.lang.IllegalArgumentException: Children: duplicate children added: parent = VBox@2d3407ae
	at javafx.graphics/javafx.scene.Parent$3.onProposedChange(Parent.java:558)
	at javafx.base/com.sun.javafx.collections.VetoableListDecorator.addAll(VetoableListDecorator.java:234)
	at javafx.base/com.sun.javafx.collections.VetoableListDecorator.addAll(VetoableListDecorator.java:103)
	at au.gov.asd.tac.constellation.views.notes.NotesViewPane.lambda$createNote$28(NotesViewPane.java:834)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at javafx.base/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
	at javafx.graphics/javafx.scene.Node.fireEvent(Node.java:8879)
	at javafx.controls/javafx.scene.control.Button.fire(Button.java:200)
	at javafx.controls/com.sun.javafx.scene.control.behavior.ButtonBehavior.mouseReleased(ButtonBehavior.java:206)
	at javafx.controls/com.sun.javafx.scene.control.inputmap.InputMap.handle(InputMap.java:274)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at javafx.base/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
	at javafx.graphics/javafx.scene.Scene$MouseHandler.process(Scene.java:3851)
	at javafx.graphics/javafx.scene.Scene$MouseHandler.access$1200(Scene.java:3579)
	at javafx.graphics/javafx.scene.Scene.processMouseEvent(Scene.java:1849)
	at javafx.graphics/javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2588)
	at javafx.graphics/com.sun.javafx.tk.quantum.EmbeddedScene.lambda$mouseEvent$4(EmbeddedScene.java:286)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at javafx.graphics/com.sun.javafx.tk.quantum.EmbeddedScene.lambda$mouseEvent$5(EmbeddedScene.java:279)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
[catch] at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:174)
	at java.base/java.lang.Thread.run(Thread.java:834)

```

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]
I think the colour picker should be part of the "Edit" mode for the Note, rather than having it on display for every Note.

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 4]
```diff
@@ -658,15 +668,54 @@ private void createNote(final NotesViewEntry newNote) {
...
+        if (newNote.getEditMode()) {
+            noteButtons = new VBox(DEFAULT_SPACING, saveTextButton, cancelButton, colourPicker);
+            newNote.setEditMode(true);
+        } else {
+            newNote.setEditMode(false);
+            noteButtons = new VBox(DEFAULT_SPACING, editTextButton, deleteButton);
+        }
+
+
+        //noteButtons.getChildren().add(colourPicker);
```
Comment (Delphinus8821) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:697]: Can we remove some of these extra blank lines

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 5]
```diff
@@ -658,15 +668,49 @@ private void createNote(final NotesViewEntry newNote) {
...
+            notesListVBox.getChildren().add(noteBody);
+        } else {
+            noteBody.setStyle("-fx-padding: 5px; -fx-background-color: "
+                    + AUTO_COLOR + "; -fx-background-radius: 10 10 10 10;");
+            notesListVBox.getChildren().add(noteBody);
+        }
+
+        // Change colour of note to whatever user sleects
+        colourPicker.setOnAction(event -> {
+            Color col = colourPicker.getValue();
```
Comment (OrionsGuardian) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:709]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 6]
```diff
@@ -787,12 +831,25 @@ private void createNote(final NotesViewEntry newNote) {
 
         // Edit button activates editable text boxs for title and label
         editTextButton.setOnAction(event -> {
+            String currentColour = newNote.getNodeColour();
```
Comment (OrionsGuardian) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:834]: final

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

**Comment by OrionsGuardian** [comment]
In the Edit Mode screen for the Note, can you move the Colour-Picker button to be above the Save and Cancel buttons, and have a larger gap between it and the save/cancel buttons. Pressing Save or Cancel will end the edit session, whereas changing the colour still keeps you in Edit mode.

**Comment by OrionsGuardian** [comment]
image
Notes on Pre-existing graphs are not being loaded properly. 
They are not being linked correctly to the corresponding nodes & transactions.

**Comment by Delphinus8821** [comment]
I think this is looking good to go once those couple of code smells have been fixed.

**Review by Delphinus8821 — APPROVED** [review]

**Review by OrionsGuardian — APPROVED** [review]

**Pull request merged**